### ISSUE TITLE
Feature Request (and code for it) - returning filepaths in ImageFolder along with data and target

### ISSUE THREAD

**Issue by Spandan-Madan** [pytorch/pytorch#2890]
To make any sense of how your model is doing, it is extremely important to visualize the results. It's great to have access to the file path along with the data and target label. 

The workaround I'm using basically overrides the getitem function by explicitly setting it to this -

```
def getitemnew(self, index):
        """
        Args:
            index (int): Index

        Returns:
            tuple: (image, target) where target is class_index of the target class.
        """
        path, target = self.imgs[index]
        img = self.loader(path)
        if self.transform is not None:
            img = self.transform(img)
        if self.target_transform is not None:
            target = self.target_transform(target)

        return img, target, path

```

I suggest adding an optional flag to ImageFolder which will allow people to get access to the paths.

**Comment by rwightman** [comment]
My instinct when I first had this need was to do it this way, but since then I've generally settled on writing side accessor functions for metadata not related to the input-target tensors for model training or evaluation. You can still do this and use the provided ImageFolder class by inheriting or wrapping and adding the accessors.

Ie in this case, you can provide a filename accessor based on index that you call if you need those filenames...

Example:
https://github.com/rwightman/pytorch-nips2017-attack-example/blob/master/dataset.py#L81

**Comment by apaszke** [comment]
I don't think adding a flag is a good idea. Subclassing ImageFolder seems ok. Also, please keep torchvision issues in pytorch/vision repo, not in the core. Thanks!
